### PR TITLE
GH-95245: Document use of `MANAGED` flags instead of offsets.

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-95245

### CODE REVIEW

File: Doc/c-api/structures.rst [hunk 1]
```diff
@@ -469,19 +469,21 @@ Accessing attributes of extension types
...
 
       static PyMemberDef spam_type_members[] = {
-          {"__dictoffset__", T_PYSSIZET, offsetof(Spam_object, dict), READONLY},
+          {"__vectorcalloffset__", T_PYSSIZET, offsetof(Spam_object, vectorcall), READONLY},
           {NULL}  /* Sentinel */
       };
 
+   The legacy offsets :c:member:`~PyTypeObject.tp_dictoffset` and
+   :c:member:`~PyTypeObject.tp_weaklistoffset` are still supported, but extensions are
+   strongly encouraged to use ``Py_TPFLAGS_MANAGED_DICT`` and
```
Comment (mdboom) [inline Doc/c-api/structures.rst:484]: Can `Py_TPFLAGS_MANAGED_DICT` and `Py_TPFLAGS_MANAGED_WEAKREF` link to their definition in `typeobj.rst` below?

File: Doc/c-api/typeobj.rst [hunk 2]
```diff
@@ -96,7 +96,7 @@ Quick Reference
    |                                                |                                   | __gt__,           |   |   |   |   |
    |                                                |                                   | __ge__            |   |   |   |   |
    +------------------------------------------------+-----------------------------------+-------------------+---+---+---+---+
-   | :c:member:`~PyTypeObject.tp_weaklistoffset`    | :c:type:`Py_ssize_t`              |                   |   | X |   | ? |
+   | (:c:member:`~PyTypeObject.tp_weaklistoffset`)  | :c:type:`Py_ssize_t`              |                   |   | X |   | ? |
```
Comment (mdboom) [inline Doc/c-api/typeobj.rst:99]: The parentheses indicate deprecation?  Should we maybe say "(Deprecated in Python 3.12)" instead?

File: Doc/c-api/typeobj.rst [hunk 3]
```diff
@@ -117,7 +117,7 @@ Quick Reference
    | :c:member:`~PyTypeObject.tp_descr_set`         | :c:type:`descrsetfunc`            | __set__,          |   |   |   | X |
    |                                                |                                   | __delete__        |   |   |   |   |
    +------------------------------------------------+-----------------------------------+-------------------+---+---+---+---+
-   | :c:member:`~PyTypeObject.tp_dictoffset`        | :c:type:`Py_ssize_t`              |                   |   | X |   | ? |
+   | (:c:member:`~PyTypeObject.tp_dictoffset`)      | :c:type:`Py_ssize_t`              |                   |   | X |   | ? |
```
Comment (mdboom) [inline Doc/c-api/typeobj.rst:120]: Same as above regarding deprecation.

File: Doc/c-api/typeobj.rst [hunk 4]
```diff
@@ -1703,48 +1731,34 @@ and :c:type:`PyType_Type` effectively act as defaults.)
...
-   specifies the offset from the *end* of the instance structure.  A negative
-   offset is more expensive to use, and should only be used when the instance
-   structure contains a variable-length part.  This is used for example to add an
-   instance variable dictionary to subtypes of :class:`str` or :class:`tuple`. Note
-   that the :c:member:`~PyTypeObject.tp_basicsize` field should account for the dictionary added to
-   the end in that case, even though the dictionary is not included in the basic
-   object layout.  On a system with a pointer size of 4 bytes,
-   :c:member:`~PyTypeObject.tp_dictoffset` should be set to ``-4`` to indicate that the dictionary is
-   at the very end of the structure.
+   The value specifies the offset from the start of the instance structure.
```
Comment (mdboom) [inline Doc/c-api/typeobj.rst]: The offset of what?  A `PyObject *` to the dictionary?

File: Doc/extending/newtypes.rst [hunk 5]
```diff
@@ -570,33 +570,20 @@ performance-critical objects (such as numbers).
...
        PyVarObject_HEAD_INIT(NULL, 0)
        /* ... other members omitted for brevity ... */
-       .tp_weaklistoffset = offsetof(TrivialObject, weakreflist),
+       .tp_flags = Py_TPFLAGS_MANAGED_WEAKREF | ...,
    };
 
+
 The only further addition is that ``tp_dealloc`` needs to clear any weak
 references (by calling :c:func:`PyObject_ClearWeakRefs`) if the field is
 non-``NULL``::
```
Comment (mdboom) [inline Doc/extending/newtypes.rst]: I think in this sentence "the field" is the `weakreflist` in the original example.  I don't think there needs to be a NULL check anymore, so maybe just remove that part of the sentence.

**Review by mdboom — CHANGES_REQUESTED** [review]

File: Doc/c-api/structures.rst (hunk 1 above)
Comment (markshannon) [reply Doc/c-api/structures.rst:484]: `:c:member:`~PyTypeObject.tp_dictoffset` converts to a link to
https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dictoffset
I don't think I can link into the table.

File: Doc/c-api/typeobj.rst (hunk 2 above)
Comment (markshannon) [reply Doc/c-api/typeobj.rst:99]: The docs are always relevant to the current version, so it is implicit that this applies to 3.12.
This isn't a formal deprecation, more of a hint not to use these fields.

**Pull request merged**